freqtrade/freqai/base_models/PyTorchModelTrainer.py

```diff
@@ -0,0 +1,184 @@
+         - Calculates the predicted output for the batch using the PyTorch model.
+         - Calculates the loss between the predicted and actual output using a loss function.
+         - Computes the gradients of the loss with respect to the model's parameters using
+           backpropagation.
+         - Updates the model's parameters using an optimizer.
+        """
+        data_loaders_dictionary = self.create_data_loaders_dictionary(data_dictionary)
+        epochs = self.calc_n_epochs(
+            n_obs=len(data_dictionary["train_features"]),
+            batch_size=self.batch_size,
+            n_iters=self.max_iters
+        )
+        for epoch in range(epochs):
+            # evaluation
+            losses = self.estimate_loss(data_loaders_dictionary, data_dictionary)
+            logger.info(
+                f"epoch ({epoch}/{epochs}):"
+                f" train loss {losses['train']:.4f} ; test loss {losses['test']:.4f}"
+            )
+            # training
+            for batch_data in data_loaders_dictionary["train"]:
+                xb, yb = batch_data
+                xb = xb.to(self.device)
+                yb = yb.to(self.device)
+                yb_pred = self.model(xb)
+                loss = self.criterion(yb_pred, yb)
+
+                self.optimizer.zero_grad(set_to_none=True)
+                loss.backward()
+                self.optimizer.step()
+
+    @torch.no_grad()
+    def estimate_loss(
+            self,
+            data_loader_dictionary: Dict[str, DataLoader],
+            data_dictionary: Dict[str, pd.DataFrame]
+    ) -> Dict[str, float]:
+
+        self.model.eval()
+        epochs = self.calc_n_epochs(
+            n_obs=len(data_dictionary["test_features"]),
+            batch_size=self.batch_size,
+            n_iters=self.eval_iters
+        )
+        loss_dictionary = {}
+        for split in ["train", "test"]:
+            losses = torch.zeros(epochs)
+            for i, batch in enumerate(data_loader_dictionary[split]):
+                xb, yb = batch
+                xb = xb.to(self.device)
+                yb = yb.to(self.device)
+                yb_pred = self.model(xb)
+                loss = self.criterion(yb_pred, yb)
+                losses[i] = loss.item()
+
+            loss_dictionary[split] = losses.mean().item()
+
+        self.model.train()
+        return loss_dictionary
+
+    def create_data_loaders_dictionary(
+            self,
+            data_dictionary: Dict[str, pd.DataFrame]
+    ) -> Dict[str, DataLoader]:
+        """
+        Converts the input data to PyTorch tensors using a data loader.
+        """
+        data_loader_dictionary = {}
+        for split in ["train", "test"]:
+            labels_shape = data_dictionary[f"{split}_labels"].shape
+            labels_view = labels_shape[0] if labels_shape[1] == 1 else labels_shape
+            dataset = TensorDataset(
+                torch.from_numpy(data_dictionary[f"{split}_features"].values).float(),
```

Comment:
from_numpy() automatically inherits input array dtype. On the other hand, torch.Tensor is an alias for torch.FloatTensor.

Therefore, if you pass int64 array to torch.Tensor, output tensor is float tensor and they wouldn't share the storage. torch.from_numpy gives you torch.LongTensor as expected.

https://stackoverflow.com/questions/48482787/pytorch-memory-model-torch-from-numpy-vs-torch-tensor

In PyTorchClassifierMultiTarget we convert the label column type from str to int. So i think it's better this way, you can control the data types outside PytorchModelTrainer.

Reply:
But `from_numpy` requires a numpy array. It is great for numpy arrays because it avoids making a copy. The point about "sharing storage" in the link you sent, only applies if you are starting from a numpy array. 

However, we are not starting with a numpy array, we are starting with a pandas dataframe, so it will convert from dataframe.values into a numpy array, and then from numpy array to torch tensor. 

`torch.tensor` skips the creation of the numpy array. 

But I have not timed it, and I doubt it is incredibly impactful since it is once per train. So do as you wish :-)

Reply:
I don't think there's a way to convert directly from DataFrame to pytorch tensor.

Reply:
There is not. There are only ways to avoid creating numpy arrays, but it is not impactful since it is once per train - either way is fine.